File: wnso-kim/Week_21/PGS_산모양타일링/PGS_산모양타일링.java
class Solution {
    public int solution(int n, int[] tops) {
        int answer = 0;
        
        int MOD = 10007;
        int[][] dp = new int[n+1][5]; // 삼각, 왼마름모, 오른마름모, 위마름모, 오위마름모
        
        dp[0][0] = dp[0][2] = 1;

        for (int i = 1; i <= n; i++) {
            dp[i][0] = (dp[i-1][0] + dp[i-1][1] + dp[i-1][2] + dp[i-1][3] + dp[i-1][4]) % MOD;
            dp[i][1] = (dp[i-1][0] + dp[i-1][1] + dp[i-1][3]) % MOD;
            
            if (i != n) {
                dp[i][2] = (dp[i-1][0] + dp[i-1][1] + dp[i-1][2] + dp[i-1][3] + dp[i-1][4]) % MOD;
            }
            
            if (tops[i-1] == 1) {
                dp[i][3] = (dp[i-1][0] + dp[i-1][1] + dp[i-1][3]) % MOD;
                
                if (i != n) {
                    dp[i][4] = (dp[i-1][0] + dp[i-1][1] + dp[i-1][3]) % MOD;
                }
            }
        }

        for (int i = 0; i < 5; i++) {
            answer = (answer + dp[n][i]) % MOD;
        }
        
        return answer;
    }
}

File: ezeun/Week_21/PGS_산모양타일링/PGS_산모양타일링.cpp
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

const int mod = 10007;

int dp[100001][2]; 
//dp[i][0]: i번째까지 정삼각or왼쪽사다리꼴or마름모로 채운 경우
//dp[i][1]: i번째까지 오른쪽사다리꼴로 채운 경우
int solution(int n, int tops[], size_t tops_len) {
    
    if(tops[0]){
        dp[0][0]=3;
        dp[0][1]=1;
    }
    else{
        dp[0][0]=2;
        dp[0][1]=1;
    }
    
    for(int i=1; i<n; i++){
        dp[i][1] = (dp[i-1][0] + dp[i-1][1])%mod;
        
        if(tops[i]){
            dp[i][0] = (dp[i-1][0]*3 + dp[i-1][1]*2)%mod;
        }
        else{
            dp[i][0] = (dp[i-1][0]*2 + dp[i-1][1])%mod;
        }
    }
    
    return (dp[n-1][0]+dp[n-1][1])%mod;
}